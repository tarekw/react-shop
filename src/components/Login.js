import React, { useState } from "react";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";

import { authenticate } from "../services/login";
import { Container } from "react-bootstrap";

export const Login = ({ setCurrentUser = (f) => f }) => {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState(false);

  const handleLogin = async (e) => {
    e.preventDefault();
    try {
      const response = await authenticate(username, password);
      setError(false);
      setCurrentUser(response.data.token);
    } catch (err) {
      console.log(err);
      setError(true);
    }
  };

  return (
    <Container fluid="sm">
      <Form onSubmit={handleLogin}>
        <Form.Group as={Row} className="mb-3">
          <Form.Label column sm="2">
            Username
          </Form.Label>
          <Col sm="10">
            <Form.Control
              type="text"
              placeholder="Enter username"
              onChange={(e) => setUsername(e.target.value)}
              isInvalid={error}
            />
          </Col>
        </Form.Group>
        <Form.Group as={Row} className="mb-3" controlId="formBasicPassword">
          <Form.Label column sm="2">
            Password
          </Form.Label>
          <Col sm="10">
            <Form.Control
              type="password"
              placeholder="Password"
              onChange={(e) => setPassword(e.target.value)}
              isInvalid={error}
            />
          </Col>
        </Form.Group>
        <Button variant="primary" type="submit">
          Sign In
        </Button>
      </Form>
    </Container>
  );
};
