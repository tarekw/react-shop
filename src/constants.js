export const BASE_URL = "https://fakestoreapi.com";
export const ALL_PRODUCTS = "All";
export const reducerTypes = {
  ADD_TO_CART: "addToCart",
  REMOVE_FROM_CART: "removeFromCart",
  SET_TOKEN: "setToken",
};
