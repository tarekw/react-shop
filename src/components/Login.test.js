/* eslint-disable  testing-library/no-wait-for-multiple-assertions */

import React from "react";
import axios from "axios";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import { waitFor } from "@testing-library/react";

import Form from "react-bootstrap/Form";
// import Button from "react-bootstrap/Button";

import { BASE_URL } from "../constants";

import { Login } from "./Login";

jest.mock("axios");

describe("Login", () => {
  beforeEach(() => {
    jest.resetAllMocks();
    axios.post.mockResolvedValue({
      data: {
        token: "1234",
      },
    });
    jest.spyOn(React, "useContext").mockReturnValue(() => {});
  });

  it("should match snapshot", () => {
    const tree = renderer.create(<Login />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should handle login", async () => {
    const wrapper = mount(<Login />);

    wrapper.find(Form.Control).at(0).getDOMNode().value = "username";
    wrapper.find(Form.Control).at(1).getDOMNode().value = "password";
    wrapper.find(Form).at(0).simulate("submit");

    await waitFor(() => {
      expect(axios.post).toHaveBeenCalledTimes(1);
      expect(axios.post).toHaveBeenCalledWith(`${BASE_URL}/auth/login`, {
        username: "username",
        password: "password",
      });
    });
  });
});
