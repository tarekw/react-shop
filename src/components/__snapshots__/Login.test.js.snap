// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match snapshot 1`] = `
<div
  className="container-sm"
>
  <form
    className=""
    onSubmit={[Function]}
  >
    <div
      className="mb-3 row"
    >
      <label
        className="form-label col-form-label col-sm-2"
      >
        Username
      </label>
      <div
        className="col-sm-10"
      >
        <input
          className="form-control"
          placeholder="Enter username"
          type="text"
        />
      </div>
    </div>
    <div
      className="mb-3 row"
    >
      <label
        className="form-label col-form-label col-sm-2"
        htmlFor="formBasicPassword"
      >
        Password
      </label>
      <div
        className="col-sm-10"
      >
        <input
          className="form-control"
          id="formBasicPassword"
          placeholder="Password"
          type="password"
        />
      </div>
    </div>
    <button
      className="btn btn-primary"
      disabled={false}
      type="submit"
    >
      Sign In
    </button>
  </form>
</div>
`;
